### PR TITLE
[TypeScript] Fix and improve type definition

### PR DESCRIPTION
**Pull request by KSXGitHub**
### Fix broken type definition

I find some type such as `combinations` and `permutations` return `Iterable<T>` (array of scalars) instead of `Iterable<T[]>` (array of arrays). This makes no sense to me, so I decided to fix it.

### Improve type definition

TypeScript 3.0+ allows users to use tuple generic arguments which I take advantage of to have more accurate function annotations.

Function definitions I changed are: `combinations`, `combinationWithReplacement`, `permutations`, and `product`.

### Added dependencies

I added [typescript-tuple](https://github.com/ksxnodemodules/typescript-tuple) as a production dependency (which I created and maintain) to utilize its generic utilities (such as `Repeat`, `Prepend`, `Reverse`).

### Potential Breaking Changes

JavaScript users are unaffected.

On TypeScript side, this change requires TypeScript version 3.0 or later, but it's OK since TypeScript is completely useless prior to https://github.com/sithmel/iter-tools/pull/54 ~(i.e. right now)~.

**TL;DR** No breaking changes.

### Additional Changes

This pull request includes changes from https://github.com/sithmel/iter-tools/pull/54 because any change of type definition without type definition being detected is useless.

### Additional Requests

1. Is it OK if I add myself as a CODEOWNER of `*.ts` files? Because I want to make sure no one give a broken type definition.

2. If this were to be merged, I ask you to merge to a "experiment" branch and publish it under "rc" tag (so that I can install it) in npm so that I can do more testing in order for it to work.

### CODE REVIEW

**Comment by KSXGitHub** [comment]
@sithmel Wait! Why did you merge it directly into master, I asked you to make an "experiment" branch and publish it under "rc" tag so that I can test it better.

**Comment by sithmel** [comment]
@KSXGitHub given your explanation about "Potential Breaking Changes" I would publish a patch release. Is that ok?
About codeowner: for now it is not necessary, given the number of contributors.
Can you add yourself to the contributors in package.json ? I think you earned it :wink:

**Comment by sithmel** [comment]
@KSXGitHub where did you ask for that ?

**Comment by KSXGitHub** [comment]
> About codeowner: for now it is not necessary, given the number of contributors.

It wasn't about credit. I want to make sure no one breaks TypeScript definition. Before the merging of this pull request, iter-tools has some incorrect type definition.

**Comment by KSXGitHub** [comment]
BTW, since you don't add me to codeowners, can you please notify me when someone change typescript files?

**Comment by sithmel** [comment]
@KSXGitHub it will be a pleasure

**Comment by sithmel** [comment]
@KSXGitHub before publishing: can you please have a look at typescript linting?
The project uses tslint and we are getting a few warnings now.
```
npm run tslint
```
If you don't like tslint or you know better alternatives, any input would be helpful. I added it, but I am not a TS user so I don't know the ecosystem very well.

**Comment by KSXGitHub** [comment]
@sithmel 

> The project uses tslint and we are getting a few warnings now.

It gives me warnings (in VSCode) even before I touch the code so I disabled it. But if you want me to fix it, I will. Is it OK if I add tslint to build script?

**Comment by sithmel** [comment]
yes please

**Comment by KSXGitHub** [comment]
Great! I'm gonna test them.
